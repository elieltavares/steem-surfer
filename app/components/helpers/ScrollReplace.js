import React from 'react';

import PropTypes from 'prop-types';

const setPathPos = (path, pos) => {
  if (window.scrollDb === undefined) {
    window.scrollDb = {};
  }

  window.scrollDb[path] = pos;
};

const getPathPos = path => {
  if (window.scrollDb) {
    return window.scrollDb[path];
  }

  return undefined;
};

class ScrollReplace extends React.Component {
  constructor(props) {
    super(props);

    this.saveTimer = null;
  }

  componentDidMount() {
    const {selector} = this.props;

    this.el = document.querySelector(selector);
    if (this.el) {
      this.el.addEventListener('scroll', this.handleScroll);
    }

    // Initial replace
    const {history, location} = this.props;
    let pos = 0;
    if (history.action === 'POP') {
      pos = getPathPos(location.pathname) || 0;
    }

    if (this.el) {
      this.el.scrollTop = pos;
    }
  }

  componentWillReceiveProps(nextProps) {
    const {location: actual} = this.props;
    const {location: next} = nextProps;

    // Replace when location change
    if (next !== actual) {
      // when page changed
      const {history} = this.props;
      let pos = 0;
      if (history.action === 'POP') {
        // if back or forward button clicked get last scroll position for location
        pos = getPathPos(next.pathname) || 0;
      }

      this.el.scrollTop = pos;
    }
  }

  componentWillUnmount() {
    this.el.removeEventListener('scroll', this.handleScroll);
  }

  handleScroll = () => {
    const {location, onBottom} = this.props;
    const pos = this.el.scrollTop;

    const save = () => {
      setPathPos(location.pathname, pos);
      // console.log(`${location.pathname} saved ${pos}`)
    };

    if (this.saveTimer) {
      clearTimeout(this.saveTimer);
    }

    this.saveTimer = setTimeout(save, 300);

    if (
      this.el.scrollTop + this.el.offsetHeight + 100 >=
      this.el.scrollHeight
    ) {
      onBottom();
    }
  };

  render() {
    return null;
  }
}


ScrollReplace.defaultProps = {
  onBottom: () => {
  }
};


ScrollReplace.propTypes = {
  location: PropTypes.shape({
    pathname: PropTypes.string.isRequired
  }).isRequired,
  history: PropTypes.shape({
    action: PropTypes.string.isRequired
  }).isRequired,
  selector: PropTypes.string.isRequired,
  onBottom: PropTypes.func
};

export default ScrollReplace;
