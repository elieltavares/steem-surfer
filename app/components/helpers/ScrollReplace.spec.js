/* eslint-disable */
import { spy } from 'sinon';
import React from 'react';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import renderer from 'react-test-renderer';
import ScrollReplace from './ScrollReplace';

Enzyme.configure({ adapter: new Adapter() });

describe('ScrollReplace', () => {
  it('(1) Render', () => {
    const props = {
      location: { pathname: '/' },
      history: { action: 'PUSH' },
      selector: '#app-content'
    };
    const component = shallow(<ScrollReplace {...props} />);
    const tree = renderer.create(component).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
