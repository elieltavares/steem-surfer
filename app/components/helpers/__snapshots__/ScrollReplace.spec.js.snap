// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ScrollReplace (1) Render 1`] = `null`;
